### ISSUE TITLE
Question about derivation of gradients of loss w.r.t. C and E

### ISSUE THREAD

**Issue by tayweekiam** [apple/ml-cross-entropy#21]
Hi, many thanks to the team for coming up with this optimisation!

May I ask how are these derived?

![Image](https://github.com/user-attachments/assets/4cc8be13-f6a1-4ac3-832b-a7815ce0b08d)

Also, how is the gradient of loss w.r.t. LSE calculated?

![Image](https://github.com/user-attachments/assets/2041198c-9c9d-48f0-aa31-91c1fcd13c24)

I am confused because it seems like the gradient of loss w.r.t. LSE is not used in the code to calculate the gradients.

Are those equations used in the paper so that the illustration of the idea can be more concise?

Thank you!

**Comment by erikwijmans** [comment]
The derivation mostly comes from the fact that the derivative of logsumexp is softmax. It's easiest to see by taking the derivative for the `i`th element in a vector `x`:
```
d/dx_i logsumexp(x) = d/dx_i log(sum(exp(x)) 
                                    = [d/dx_i sum(exp(x))] / sum(exp(x)) # chain rule for log(x)
                                    = d/dx_i exp(x_i) / sum(exp(x)) # expand the sum and select the ith element in the sum
                                    = exp(x_i) / sum(exp(x)) [d/dx_i x_i] # chain rule for exp(x)
                                    = exp(x_i) / sum(exp(x)) [d/dx_i x_i] # d/dx_i x_i = 1
                                    = softmax(x) # simplify
```



Given this, $\nabla E$ is as follows
```
dE = d/dE log(sum(exp(C^T E)))
     = softmax(C^T E) (d/dE C^T E) # chain rule using d/dx logsumexp(x) = softmax(x)
     = softmax(C^T E) C # d/dE C^T E = C
```
The derivation for $\nabla C$ follows a similar path.

$\nabla LSE$ is a quantity provided by whatever is downstream of CCE. It'll depend on how the loss per input gets combined into the mini-batch loss. In the case of averaging, this will be a vector where all elements are `1/N`.

In the code, $\nabla LSE$ is calculated by PyTorch's autograd. CCE outputs `ce = lse - correct_token_logit` and in the backward pass PyTorch will give CCE the gradient w.r.t to its output. So, given $\nabla ce$, then $\nabla LSE = \nabla ce$

**Comment by tayweekiam** [comment]
Thank you for the clarification!

Given `ce = lse - correct_token_logit`, and both `lse` and `correct_token_logit` are functions of C and E, why is $\nabla LSE = \nabla ce$?.

How I am seeing it is that this situation is similar to finding `df/d(x^2)` where `f = x^2 - x`. In this example, `df/d(x^2)` is not equal to 1.

I am quite new to this field, sorry if it's a trivial question.

**Comment by erikwijmans** [comment]
From the perspective of the `ce` function, `lse` and `correct_token_logits` are just variables. The fact that they themselves are also functions doesn't matter when computing how changes in `lse` will change the value of `ce`. Therefore `d(CE)/d(LSE) = 1`. The gradient of LSE w.r.t to the loss, `dL/d(LSE)`, is `dL/d(CE) * d(CE)/d(LSE)`. Using `d(CE)/d(LSE) = 1`, we get `dL/d(LSE) = dL/d(CE)`.


`df/d(x^2)` is 1 in your example. `df/dx` is not 1, but `f` is a linear function with slope 1 w.r.t to x^2, so `df/d(x^2)` is 1. If we instead did `f = a - x` where `a = x^2`, then `df/da` being 1 may be clearer.

**Comment by tayweekiam** [comment]
Were there some assumptions involved?

It seems to be quite different from how I understood it. I managed to find [this] (https://math.stackexchange.com/a/3458396) post that explained my thoughts on why `df/d(x^2)` (and hence `d(CE)/d(LSE)`)  is not 1.

**Comment by erikwijmans** [comment]
I am considering `x^2` to be just a variable when talking about `df/d(x^2)`, not a function of `x`. If we have `f = a + t`, then `df/da = 1` and `df/dt = 1`. We can then set `a = x^2` and `t = -x` and get `df/dx = df/da da/dx + df/dt dt/dx = 1 * 2x + 1 * (-1) = 2x - 1`, which is correct for `f = x^2 - x`.

I don't entirely get what they are doing in that stack exchange post. I think they are talking about `dy/d(x^2)` and considering `x^2` to be a function of `x`. I'm sure that can be a useful framing, but it isn't when successively applying the chain rule to do back-prop. In back-prop, we take into account that `x^2` is a function of `x` in later steps, like I did above.

**Comment by tayweekiam** [comment]
Okay I think I got it. I made a mistake when thinking about the chain rule part. Thank you for your patience!
